fish-rust/src/builtins/string/repeat.rs

```diff
@@ -0,0 +1,142 @@
+        let mut print_newline = true;
+
+        for (w, want_newline) in Arguments::new(args, optind, streams, true) {
+            print_newline = want_newline;
+            if w.is_empty() {
+                continue;
+            }
+
+            all_empty = false;
+
+            if self.quiet {
+                // Early out if we can - see #7495.
+                return STATUS_CMD_OK;
+            }
+
+            if !first {
+                streams.out.append1('\n');
+            }
+            first = false;
+
+            // The maximum size of the string is either the "max" characters,
+            // or it's the "count" repetitions, whichever ends up lower.
+            let max = if self.max == 0 || (self.count > 0 && w.len() * self.count < self.max) {
+                w.len() * self.count
+            } else {
+                self.max
+            };
+
+            // Reserve a string to avoid writing constantly.
+            // The 1500 here is a total gluteal extraction, but 500 seems to perform slightly worse.
+            let chunk_size = 1500;
+            // The + word length is so we don't have to hit the chunk size exactly,
+            // which would require us to restart in the middle of the string.
+            // E.g. imagine repeating "12345678". The first chunk is hit after a last "1234",
+            // so we would then have to restart by appending "5678", which requires a substring.
+            // So let's not bother.
+            //
+            // Unless of course we don't even print the entire word, in which case we just need max.
+            let mut chunk = WString::with_capacity(max.min(chunk_size + w.len()));
+
+            let mut i = max;
+            loop {
+                if i >= w.len() {
+                    chunk.push_utfstr(&w);
+                } else {
+                    chunk.push_utfstr(w.slice_to(i));
+                    break;
+                }
+
+                if chunk.len() >= chunk_size {
+                    // We hit the chunk size, write it repeatedly until we can't anymore.
+                    streams.out.append(&chunk);
+                    while i >= chunk.len() {
+                        streams.out.append(&chunk);
+                        // We can easily be asked to write *a lot* of data,
+                        // so we need to check every so often if the pipe has been closed.
+                        // If we didn't, running `string repeat -n LARGENUMBER foo | pv`
+                        // and pressing ctrl-c seems to hang.
+                        if streams.out.flush_and_check_error() != STATUS_CMD_OK.unwrap() {
+                            return STATUS_CMD_ERROR;
+                        }
+                        i -= chunk.len();
+                    }
+                    chunk.clear();
+                }
+
+                let Some(new_i) = i.checked_sub(w.len()) else {
+                    break
+                };
+                i = new_i;
```

Comment:
> I don't think there's the possibility of overflow here because otherwise the C++ code would have a possible infinite loop, no?
Can we keep the loop with while i > 0?

Ah yeah, that's covered by the `else` at the top

```suggestion
            while i > 0 {
                if i >= w.len() {
                    chunk.push_utfstr(&w);
                } else {
                    chunk.push_utfstr(w.slice_to(i));
                    break;
                }

                if chunk.len() >= chunk_size {
                    // We hit the chunk size, write it repeatedly until we can't anymore.
                    streams.out.append(&chunk);
                    while i >= chunk.len() {
                        streams.out.append(&chunk);
                        // We can easily be asked to write *a lot* of data,
                        // so we need to check every so often if the pipe has been closed.
                        // If we didn't, running `string repeat -n LARGENUMBER foo | pv`
                        // and pressing ctrl-c seems to hang.
                        if streams.out.flush_and_check_error() != STATUS_CMD_OK.unwrap() {
                            return STATUS_CMD_ERROR;
                        }
                        i -= chunk.len();
                    }
                    chunk.clear();
                }

                i -= w.len();
```

Reply:
github doesn't support ignoring whitespace in suggestions, I'll take a look with `diff -w`